### DISCUSSION TITLE
Which DFS Should We Use?

### DISCUSSION THREAD

**Discussion by waterflier** [buckyos/buckyos#169] [General]
Hi everyone,

As part of our plan for the next Alpha release of BuckyOS (Alpha2), we aim to integrate a Distributed File System (DFS). In the previous demo version, we used GlusterFS. 

Could GlusterFS still be the answer?

The choice of DFS is crucial, as it directly impacts both the safety of user data and the foundational performance of the system. I believe it’s important for us to revisit this decision with a focus on our "Final Goals," and ideally, support our choice with rational data. Below, I’ve outlined the core requirements I’ve considered:

1. **Target Audience and Deployment Model**:  
   BuckyOS is designed for regular users. The envisioned storage setup doesn’t involve dynamic addition or replacement of hard drives; instead, all storage devices are pre-installed and configured during manufacturing to operate at optimal performance. When users need more storage space, they are expected to add a new storage device (not storage disk) within their Zone (e.g., purchasing another unit). This is a key reason we require a DFS rather than a RAID system.  

2. **Cluster Size Considerations**:  
   Practically speaking, most home users will likely have 1-3 storage devices, while small business users may have 3-15. This means we are dealing with small-scale clusters, especially clusters with fewer than 5 devices, which differ significantly from the scenarios most DFS systems are designed for.  

3. **Single Device Scenario**:  
   For a single storage device, the typical setup would include:  
   - Four HDDs (same capacity) for storage  
   - One SSD for system installation and read/write caching  
   Data is stored across the four HDDs using `erasure coding`, ensuring that failure of any single HDD does not result in data loss (although two HDD failures all). The effective capacity is **3× the capacity of a single HDD**.  

4. **Small Cluster (3 OOD) Setup**:  
   In our roadmap, users will quickly transition to a standard 3 OOD (Owner Online Device) cluster. This consists of:  
   - One storage device  
   - One StorageBox  
   - One VPS  
   In this setup, the system can tolerate the unavailability of any single OOD.  

5. **Handling Storage Device Failures**:  
   If the unavailable OOD is the storage device, the DFS service will down:  
   - **Read Capability**: The system can only access cached hot data.  
   - **Write Capability**: In principle, writes are still possible until the reliable write cache is full (this feature is typically restricted to social-network-related apps).  

6. **Adding a Second Storage Device**:  
   Users may expand the cluster by adding a second storage device with, for example, 6 HDDs. According to the aforementioned consensus, the effective capacity of this device is **5× the capacity of a single HDD**. The system will then see a total effective storage capacity of **8× a single HDD’s capacity (3+5)**.  

7. **Partition vs. Distributed Storage Design (Discussion Needed)**:  
   When two "independent" storage devices exist in the cluster, the failure of either device would render some partitions (directories) unavailable—similar to the situation with single hard drive failure on a PC. While this design does not leverage DFS’s full potential for balancing storage capacity, distributing every app's storage across both devices would mean that failure of one device results in incomplete storage for all apps.  

8. **Reaching the "DFS Moment"**:  
   Over time, users will expand to 3 (or 5?) storage devices, marking the "DFS moment." A killer-App that allows users to set up 3 devices simultaneously would be ideal. Otherwise, natural expansion rates suggest this milestone might take around three years to achieve. Once the DFS moment is reached, the DFS can transition to its conventional mode, allowing any single storage device to go offline while ensuring all data remains intact and accessible. For example, if each of the 3 storage devices has 4 HDDs, the effective capacity would be **8× a single HDD’s capacity**. Before reaching the DFS moment, the capacity would have been **9× a single HDD’s capacity**—the user sacrifices some space in exchange for improved availability.  

---

### Some Aspirations for Advanced Users:  

For advanced users, I hope it will be as simple as running software on an existing Home Server/NAS to contribute part of its storage to the DFS. For instance, my workstation already has 24TB of capacity, and I’d like to allocate 10TB to BuckyOS.  

---

Looking forward to hearing your thoughts!

**Comment by photosssa** [comment]
I came across this project: https://github.com/seaweedfs/seaweedfs. It is implemented in Go and has a significant number of stars on GitHub, with frequent development commits and active discussions. The structure and implementation are straightforward. It supports a topology with three levels: data center (dc), rack, and node. It features a 2n+1 master HA implementation supported by the Raft protocol, as well as disk configurations for heterogeneous storage (such as SSDs, HDDs, etc.) with labels. It separates optional S3 and FUSE interfaces from the filer service, and supports data HA through two methods: replication and offline erasure coding (RS).You might consider using this project to implement a DFS (Distributed File System).

**Comment by wugren** [comment]
glusterfs does not support ssd cache.

glusterfs expansion of the same volume only supports devices of the same type, and does not support expansion of different storage spaces and the number of storage disks.
